### PR TITLE
Refactor freedom-ldscript-generator

### PR DESCRIPTION
**Pull request by nategraff-sifive**
The central idea of this refactor is to capture the "strategy"
abstraction. Each strategy knows which set of memories must
be present in the devicetree for it to be valid, and then the ability to
decide which memory to use for each purpose (ram, rom, itim, etc) when
generating a linker script.

To extend the functionality of the linker script to a new layout, create
a new strategy.

The ChosenStrategy strategy is also created which allows the `chosen`
DeviceTree node to specify which node to use for each purpose using the
 - `metal,ram`
 - `metal,rom`
 - `metal,itim`

properties.

### CODE REVIEW

**Comment by nategraff-sifive** [comment]
I switched automake to use tar-ustar instead of the legacy default tar-v7 to allow filenames up to 256 bytes long. This should (?) still be portable enough for our needs, but 99 bytes was getting really limiting to try and keep to that limit.

**Comment by nategraff-sifive** [comment]
Tested works with hello on all standard and spanning configs.

File: linker_script/memory.c++ [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+{
+  return (this->base < rhs.base);
+}
+
+string Memory::describe()
+{
+  stringstream stream;
+
+  stream << hex << name << " (" << attributes << ") : ORIGIN = 0x" << base
+         << ", LENGTH = 0x" << size;
```
Comment (bsousi5) [inline linker_script/memory.c++]: We are not longer limiting the memory size 2GB or below.
I think we should keeps that.
We should also add check to to make sure the total memory map is below 2GB, now that we have a mean to sort the complete memory map.

Comment (nategraff-sifive) [reply linker_script/memory.c++]: I'll add this check in a subsequent PR.

**Pull request merged**